clang/lib/Sema/SemaFunctionEffects.cpp

```diff
@@ -627,7 +627,7 @@ class Analyzer {
           IsNoexcept = isNoexcept(FD);
         } else if (auto *BD = dyn_cast<BlockDecl>(D)) {
           if (auto *TSI = BD->getSignatureAsWritten()) {
-            auto *FPT = TSI->getType()->getAs<FunctionProtoType>();
+            auto *FPT = TSI->getType()->castAs<FunctionProtoType>();
```

Comment:
I'm not sure that if we have a signature, we're guaranteed to have the FunctionProtoType. Perhaps we should leave `getAs` and check that returned value is not null before working with it.
@AaronBallman , any change you can help to clarify that?

Reply:
This code path is hit from:
```
  if (Context.hasAnyFunctionEffects())
    performFunctionEffectAnalysis(Context.getTranslationUnitDecl());
```
and the only thing that sets `AnyFunctionEffects` is through a call to `ASTContext::getFunctionTypeInternal()`, which is only called for functions with prototypes. So I believe this change is correct as-is, but is a bit fragile.

CC @Sirraide @cjappl @dougsonos for extra eyes; are K&R C functions supposed to support function effects? If so, I suspect that doesn't work today.

Reply:
> are K&R C functions supposed to support function effects?

They should yes; we had to add checks for functions w/o prototypes in several places to avoid crashes.

Reply:
Well, also, we only get here if we’re analysing a block. Can blocks be prototypeless? I hope not...

Reply:
Actually, both of those questions are irrelevant because this entire code path is guarded by a check for `S.getLangOpts().CPlusPlus`. So yes, changing this to `castAs` is fine.

Reply:
I'm mobile and away from code until Monday, but IIRC effects are stored as trailing objects on FunctionProtoType and we simply don't allow them on prototype-less functions. I think there's a custom diagnostic if the user tries. The test is in something like Sema/attr-nonblocking-syntax.c.

Reply:
Right, I think what I was remembering then is that we might sometimes end up w/ prototypeless functions in the analysis code, but we just don’t do much w/ them.